### PR TITLE
[CELEBORN-227] Support different push strategies to control the push speed

### PR DESCRIPTION
**Pull request by boneanxs**
### What changes were proposed in this pull request?
1. Abstract the congest control strategy to allow different strategies.
2. For `SLOW_START` strategy, add an additional sleep time to let the speed more slower, since the push speed cannot be controlled if only restrict the maxReqs to 1.
3. The strategy should respect the target worker, such as `SLOWSTART` can provide different `currentMaxReqInFlight` for different target worker.

### Why are the changes needed?

no

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
Add tests to cover this.

### CODE REVIEW

**Comment by boneanxs** [comment]
@waitinfuture @AngersZhuuuu Please help to review this.

File: client/src/main/java/org/apache/celeborn/client/write/SlowStartPushSpeedStrategy.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    this.continueCongestedNumber = new AtomicInteger(0);
+    this.maxInFlight = conf.pushMaxReqsInFlight();
+    this.maxSleepMills = conf.pushSlowStartMaxSleepMills();
+  }
+
+  /**
+   * If `pushDataSlowStart` is enabled, will increase `currentMaxReqsInFlight` gradually to meet the
+   * max push speed.
+   *
+   * <p>1. slow start period: every RTT period, `currentMaxReqsInFlight` is multiplied.
```
Comment (waitinfuture) [inline client/src/main/java/org/apache/celeborn/client/write/SlowStartPushSpeedStrategy.java:53]: multiplied by some value?

**Comment by waitinfuture** [comment]
Thanks for the pr! @boneanxs I have left some comments, feel free to ping me anytime

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 2]
```diff
@@ -2116,6 +2116,25 @@ object CelebornConf extends Logging {
       .timeConf(TimeUnit.MILLISECONDS)
       .createOptional
 
+  val PUSH_LIMIT_STRATEGY: ConfigEntry[String] =
+    buildConf("celeborn.push.limit.strategy")
+      .categories("client")
+      .doc("The strategy used to control the push speed.")
```
[the remark this answers is not in the document]
Comment (boneanxs) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:2122]: Sure, will do

File: client/src/main/java/org/apache/celeborn/client/write/SlowStartPushSpeedStrategy.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        continueCongestedNumber.incrementAndGet();
+      } else {
+        currentMaxReqsInFlight.updateAndGet(pre -> pre / 2);
+      }
+      maxInFlight = currentMaxReqsInFlight.get();
+      congestionAvoidanceFlag = 0;
+    }
+  }
+
+  protected long getSleepTime(int currentMaxReqs) {
```
[the remark this answers is not in the document]
Comment (boneanxs) [reply client/src/main/java/org/apache/celeborn/client/write/SlowStartPushSpeedStrategy.java:92]: Yea, the main purpose is for test.

File: client/src/main/java/org/apache/celeborn/client/write/SlowStartPushSpeedStrategy.java (hunk 1 above)
Comment (boneanxs) [reply client/src/main/java/org/apache/celeborn/client/write/SlowStartPushSpeedStrategy.java:53]: will change, here means `currentMaxReqsInFlight` is doubled

File: client/src/main/java/org/apache/celeborn/client/write/SlowStartPushSpeedStrategy.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    if (currentMaxReqs >= conf.pushMaxReqsInFlight()) {
+      return 0;
+    }
+
+    long sleepInterval = 2000 - 60L * currentMaxReqs;
```
[the remark this answers is not in the document]
Comment (boneanxs) [reply client/src/main/java/org/apache/celeborn/client/write/SlowStartPushSpeedStrategy.java:97]: If the worker side doesn't congest this client, the currentMaxReqsInFlight will increase exponentially(0, 1, 2, 4, 8, 16, 32), and the max sleep time is approximately 10s

File: client/src/main/java/org/apache/celeborn/client/write/SlowStartPushSpeedStrategy.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+      if (currentMaxReqsInFlight.get() >= maxInFlight) {
+        // Congestion avoidance
+        congestionAvoidanceFlag++;
+        if (congestionAvoidanceFlag >= currentMaxReqsInFlight.get()) {
+          currentMaxReqsInFlight.incrementAndGet();
+          congestionAvoidanceFlag = 0;
+        }
+      } else {
+        // Slow start
+        currentMaxReqsInFlight.incrementAndGet();
```
[the remark this answers is not in the document]
Comment (boneanxs) [reply client/src/main/java/org/apache/celeborn/client/write/SlowStartPushSpeedStrategy.java:73]: one RTT period means one batch(currentMaxReqsInFlight) of requests, so here we only increase once for every response, and the total `currentMaxReqsInFlight` will be increased to double before.

File: client/src/main/java/org/apache/celeborn/client/write/SlowStartPushSpeedStrategy.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (waitinfuture) [reply client/src/main/java/org/apache/celeborn/client/write/SlowStartPushSpeedStrategy.java:97]: IMO 10s is too long. For many stages the task can be less than 10 seconds. Maybe we can change 2000 to 500 or something based on testing, and make it configurable.

[the remark this answers is not in the document]
Comment (boneanxs) [reply client/src/main/java/org/apache/celeborn/client/write/SlowStartPushSpeedStrategy.java:97]: Yea, make sense, give the fact that we will punish the client if it keeps being congested even the `maxReq` is 1, maybe we can reduce the 2s to a lower value, let me have a test.

[the remark this answers is not in the document]
Comment (boneanxs) [reply client/src/main/java/org/apache/celeborn/client/write/SlowStartPushSpeedStrategy.java]: Still can congest the users if make the initial sleep time smaller, now the total sleep time is 1s

```bash
#OFFHEAP_MEMORY 20G
rss.worker.memory.check.interval                5
rss.worker.memory.report.interval               5s
rss.pause.pushdata.memory.ratio                 0.9
rss.pause.replicate.memory.ratio                0.98
rss.resume.memory.ratio                         0.7

celeborn.worker.ratelimit.enabled                    true
celeborn.worker.ratelimit.sample.time.window         5s
celeborn.worker.ratelimit.low.watermark              14G
celeborn.worker.ratelimit.high.watermark             17G
celeborn.worker.ratelimit.user.inactive.interval     5min
```

Screenshot 2023-02-02 at 18 59 02
Screenshot 2023-02-02 at 19 00 18

**Comment by AngersZhuuuu** [comment]
In this pr you also support slow start and congestion control for different target host port, pls add this information in title and pr desc.

File: common/src/main/java/org/apache/celeborn/common/write/SlowStartPushStrategy.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+  protected CongestControlContext getCongestControlContextByAddress(String hostAndPushPort) {
+    return congestControlInfoPerAddress.computeIfAbsent(
+        hostAndPushPort, host -> new CongestControlContext(maxInFlight));
+  }
+
+  /**
+   * If `pushDataSlowStart` is enabled, will increase `currentMaxReqsInFlight` gradually to meet the
+   * max push speed.
+   *
+   * <p>1. slow start period: every RTT period, `currentMaxReqsInFlight` is doubled.
```
Comment (waitinfuture) [inline common/src/main/java/org/apache/celeborn/common/write/SlowStartPushStrategy.java:107]: I don't see where currentMaxReqsInFlight is doubled, should it be line 58?

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 7]
```diff
@@ -27,6 +27,7 @@ import scala.util.Try
 
 import org.apache.hadoop.security.UserGroupInformation
 
+import org.apache.celeborn.common.CelebornConf.PUSH_LIMIT_STRATEGY
```
Comment (waitinfuture) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:30]: unused import

**Review by waitinfuture — APPROVED** [review]
LGTM exception some small comments, Thanks!

File: common/src/main/java/org/apache/celeborn/common/write/SlowStartPushStrategy.java (hunk 6 above)
Comment (boneanxs) [reply common/src/main/java/org/apache/celeborn/common/write/SlowStartPushStrategy.java:107]: Yea, from @AngersZhuuuu's initial idea, here we define one RTT period means one batch of push data requests(batch size -> `currentMaxReqsInFlight`). So we plus 1 for every request, and for one batch request, the next batch will be increased to currentMaxReqsInFlight * currentMaxReqsInFlight, which is doubled.

e.g.
current currentMaxReqsInFlight is 1, the next maxReqs is 2
current currentMaxReqsInFlight is 2, the next maxReqs is 4, etc...

Comment (AngersZhuuuu) [reply common/src/main/java/org/apache/celeborn/common/write/SlowStartPushStrategy.java:107]: > I don't see where currentMaxReqsInFlight is doubled, should it be line 58?

In this method `CongestControlContext .increaseCurrentMaxReqs()`

**Review by AngersZhuuuu — APPROVED** [review]
LGTM

**Pull request merged**